sycl/include/CL/sycl/queue.hpp

```diff
@@ -754,285 +754,244 @@ class __SYCL_EXPORT queue {
-  /// \param CodeLoc contains the code location of user code
-  template <typename KernelName = detail::auto_name, typename KernelType>
-  event parallel_for(range<2> NumWorkItems, const std::vector<event> &DepEvents,
-                     _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
-    _CODELOCARG(&CodeLoc);
-    return parallel_for_impl<KernelName>(NumWorkItems, DepEvents, KernelFunc,
-                                         CodeLoc);
+  /// \param Rest acts as-if: "ReductionTypes&&... Reductions,
+  /// const KernelType &KernelFunc".
+  template <typename KernelName = detail::auto_name, typename... RestT>
+  event parallel_for(range<2> Range, const std::vector<event> &DepEvents,
+                     RestT &&...Rest) {
+    return parallel_for_impl<KernelName>(Range, DepEvents, Rest...);
   }
 
   /// parallel_for version with a kernel represented as a lambda + range that
   /// specifies global size only.
   ///
-  /// \param NumWorkItems is a range that specifies the work space of the kernel
+  /// \param Range specifies the global work space of the kernel
   /// \param DepEvents is a vector of events that specifies the kernel
   /// dependencies
-  /// \param KernelFunc is the Kernel functor or lambda
-  /// \param CodeLoc contains the code location of user code
-  template <typename KernelName = detail::auto_name, typename KernelType>
-  event parallel_for(range<3> NumWorkItems, const std::vector<event> &DepEvents,
-                     _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
-    _CODELOCARG(&CodeLoc);
-    return parallel_for_impl<KernelName>(NumWorkItems, DepEvents, KernelFunc,
-                                         CodeLoc);
+  /// \param Rest acts as-if: "ReductionTypes&&... Reductions,
+  /// const KernelType &KernelFunc".
+  template <typename KernelName = detail::auto_name, typename... RestT>
+  event parallel_for(range<3> Range, const std::vector<event> &DepEvents,
+                     RestT &&...Rest) {
+    return parallel_for_impl<KernelName>(Range, DepEvents, Rest...);
   }
 
   /// parallel_for version with a kernel represented as a lambda + range and
   /// offset that specify global size and global offset correspondingly.
   ///
-  /// \param NumWorkItems is a range that specifies the work space of the kernel
+  /// \param Range specifies the global work space of the kernel
   /// \param WorkItemOffset specifies the offset for each work item id
   /// \param KernelFunc is the Kernel functor or lambda
   /// \param CodeLoc contains the code location of user code
   template <typename KernelName = detail::auto_name, typename KernelType,
             int Dims>
-  event parallel_for(range<Dims> NumWorkItems, id<Dims> WorkItemOffset,
-                     _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
-    _CODELOCARG(&CodeLoc);
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
+  event parallel_for_impl(range<Dims> Range, id<Dims> WorkItemOffset,
```

Comment:
Yes, I did that intentionally. The main entry point for this args-configuration is the func at the line 761.
It would be ambiguous call/functions error without renaming this code here to parallel_for_impl

Reply:
> someone might still be using it

May be someone is using it, maybe not, but for sure we have 3 LIT test using parallel_for with 'offset'. 
Fails on those tests made me rename this function to parallel_for_impl.

Reply:
Ok, thank you!